### ISSUE TITLE
Toy model #1

### ISSUE THREAD

**Issue by b-remy** [b-remy/gems#1]
This issue is to document the choice of prior parameters in our model.

So far, we assume the following PGM:
![pgm-toy-model1](https://user-images.githubusercontent.com/30293694/133829145-4a08b427-c875-4418-aa6e-ae44bcecc899.png)

Thus, we need to set prior on galaxy size, shape noise and shear field.

We can use the COSMOS catalog from GalSim to estimate such parameters. In the following notebook, I fitted distributions from the COSMOS catalog and the outcomes are:

https://gist.github.com/b-remy/bd7acd3255b223004dc517462a10489d

- log10 (half_light_radius) follows N(-0.68, 0.30)
- log10 (sersic) follows N(0.1, 0.39)
- log10 (intensity) follows N(-1.97, 0.53)
- shape noise follows N(0, 0.28)

To build Sersic light profile with GalFlow, we need to specify the Flux alongside the size and sersic index. In the notebook, I show how the Flux can be computed from hlr, ellipticity and intensity.

To compute shear distribution, I averaged ellipticities by bins of 10 and fitted the histogram of binned shapes. Thus,

- shear (avg of 10 gal) follows N(0, 0.09)

It still remains to create a PSF model (that is fixed, no random variable).

According to the PGM above and the notebook fits, should we split `lambda` in three parameters `hlr`, `n` and `I` ? or should we only keep `hlr` varying?

**Comment by EiffL** [comment]
oh wow, tu cartones! 
It's looking great :-) 
when you say shear here, is it gravitational shear or intrinsic galaxy ellipticity?

**Comment by b-remy** [comment]
I actually wanted to do shearing but I didn't think a lot about it. Please tell me what you think about this:

After creating the profile, I can do a step of shearing to simulate intrinsic gal ellipticity, sampling e1, e2 from N(0, 0.28) (as estimated in the notebook above).

Then for actual shear, should I apply the same value to all the galaxies, assuming that they all belong to the same bin?
Since I have images of 5x5 stamps, I can estimate shear with averages of 25 ellipticities in the COSMOS catalog and thus following N(0., 0.05).

**Comment by EiffL** [comment]
yep, trying to infer a constant shear is not a bad first step! Maybe let's see if we can run some inference in this case to get started?

Where things become really fun though is where you try to estimate a spatially varying shear. We could for instance start with a simple model where we model the convergence as Gaussian Process, I think that's what Michael was doing. And then we could upgrade it with our score matching model :-)

**Comment by EiffL** [comment]
Just sharing a small trick for reshaping a batch of images into a 2d image ;-) 
```python
ims = model(16, 64)
res = ims.numpy().reshape(4,4,64,64).transpose([0,2,1,3]).reshape([4*64,4*64])
imshow(res, cmap='gray_r')
```
![image](https://user-images.githubusercontent.com/861591/134239318-8358dc04-5d88-4ed6-b61c-58b5e4ada05b.png)

**Comment by EiffL** [comment]
Another trick :-) You can tweak your model function so that it outputs the likelihood, which you can then directly use to compute the joint likeilhood, here is what I mean:

```python
def model(batch_size, stamp_size):
  [....]
  # Convolve the image with the PSF
  profile = galflow.convolve(ims, kpsf,
                      zero_padding_factor=padding_factor,
                      interp_factor=interp_factor)[...,0]
  # Returns likelihood
  return  ed.Normal(loc=profile, scale=sigma_e)
```
This way you can for instance do things like `model().distribution.log_pob(data)`, even better if you can conditioned the model on some specific parameter values. 

For computing the joint likelihood, with a PPL you would typically do it in 2 steps:
1. Condition the model on observed data
2. Compute the joint log likelihood.

In Edward2 check that API here: https://github.com/google/edward2/blob/e75d4db9f9a7dfedbbd0746ca64fdfdac2d0f399/edward2/tracers.py#L31

I think we can do something like:
```python
ed.make_joint_log(ed.condition(model, obs=ims))
```

**Comment by EiffL** [comment]
hummmm this conditionning things isnt working as I expected... but anyway you can do this instead:

```python
log_prob = ed.make_log_joint_fn(model)

log_prob(n=true_params['n'], 
         hlr=true_params['hlr'],
         gamma=true_params['gamma'],
         obs=ims)
```

**Comment by EiffL** [comment]
assuming you have named all variables in your forward model.

**Comment by EiffL** [comment]
Sooooo cool :-D Here is a little example of running a chain, just leaving the ellipticity as a free param, and starting chain at 0 ellipticity for all the objectst:
![image](https://user-images.githubusercontent.com/861591/134247880-a1610c28-6cd7-46c5-8e89-d79bc1aaf577.png)

Full notebook [here](https://colab.research.google.com/drive/1w2W_ISV_7sMYry7VzPFWTWk1b71AcBge?usp=sharing)

**Comment by EiffL** [comment]
But there seems to be an issue with either hlr or n if  I sample many images some of them end up with NaNs, and then in the HMC chain it blows up. It's usually the sersic index that causes issues but havent figured out exactly where these nans are coming from

**Comment by b-remy** [comment]
I noticed the same thing later today after opening https://github.com/DifferentiableUniverseInitiative/GalFlow/pull/15... I have to do more tests to figure out because I thought that I handled all the NaNs.... I also have to enable `@tf.function` using `tf.while_loop` inside `calculate_b(n)` function.

**Comment by b-remy** [comment]
But this is a very cool demonstration !!

**Comment by EiffL** [comment]
yep, just updated it to include a constant shear + intrinsic galaxy ellipticity. It's a bit slow to sample because not tf.functioned but "l'idee y est" :-)

**Comment by EiffL** [comment]
And now with chain and posterior on shear: https://colab.research.google.com/drive/1hyZBlG5KhNHF1rCx3U7BFUVOkZ75e0CR?usp=sharing

![image](https://user-images.githubusercontent.com/861591/134464488-d1add407-c6e3-405e-ad2f-802e5c9b09f5.png)


![image](https://user-images.githubusercontent.com/861591/134464458-0e8b4c38-cc8a-4752-a0a6-6ad1ced58995.png)


![image](https://user-images.githubusercontent.com/861591/134464441-082ba922-6fe4-4094-9081-1ad9d4f588d7.png)

**Comment by aguinot** [comment]
Hi there,
Very nice results!
I am wondering if you have investigate why the estimation on g2 is much noisier than g1? Do you think this is just a coincidence?

Also by checking your example I have noticed that you are using sersic profile, right? If this is the case you might consider using a "simpler" profile just for testing given that the implementation of this profile in galsim is "slow" if you change the sersic index a lot (Check this issue for more info https://github.com/GalSim-developers/GalSim/issues/566). By taking a De Vaucouleur or Exponential you might speed up your tests..
